OpenRA.Mods.Common/Activities/Air/Fly.cs

```diff
@@ -96,38 +136,38 @@ public override Activity Tick(Actor self)
 			if (useLastVisibleTarget && !lastVisibleTarget.IsValidFor(self))
 				return NextActivity;
 
-			var pos = self.CenterPosition;
 			var checkTarget = useLastVisibleTarget ? lastVisibleTarget : target;
-
-			if (!soundPlayed && aircraft.Info.TakeoffSounds.Length > 0 && self.IsAtGroundLevel())
-			{
-				Game.Sound.Play(SoundType.World, aircraft.Info.TakeoffSounds, self.World, aircraft.CenterPosition);
-				soundPlayed = true;
-			}
-
-			// We are taking off, so remove influence in ground cells.
-			if (self.IsAtGroundLevel())
-				aircraft.RemoveInfluence();
+			var delta = checkTarget.CenterPosition - self.CenterPosition;
+			var desiredFacing = !aircraft.Info.CanHover || delta.HorizontalLengthSquared != 0 ? delta.Yaw.Facing : aircraft.Facing;
 
 			// Inside the target annulus, so we're done
 			var insideMaxRange = maxRange.Length > 0 && checkTarget.IsInRange(aircraft.CenterPosition, maxRange);
 			var insideMinRange = minRange.Length > 0 && checkTarget.IsInRange(aircraft.CenterPosition, minRange);
 			if (insideMaxRange && !insideMinRange)
 				return NextActivity;
 
-			var delta = checkTarget.CenterPosition - self.CenterPosition;
+			var move = aircraft.Info.CanHover ? aircraft.FlyStep(desiredFacing) : aircraft.FlyStep(aircraft.Facing);
+
+			// Inside the minimum range, so reverse if CanHover
+			if (aircraft.Info.CanHover && insideMinRange)
+			{
+				FlyTick(self, aircraft, desiredFacing, aircraft.Info.CruiseAltitude, -move);
+				return this;
+			}
 
-			// The next move would overshoot, so consider it close enough
-			var move = aircraft.FlyStep(aircraft.Facing);
+			// The next move would overshoot, so consider it close enough or set final position if CanHover
 			if (delta.HorizontalLengthSquared < move.HorizontalLengthSquared)
+			{
+				if (aircraft.Info.CanHover)
+				{
+					var targetAltitude = self.World.Map.DistanceAboveTerrain(checkTarget.CenterPosition);
+					aircraft.SetPosition(self, checkTarget.CenterPosition + new WVec(WDist.Zero, WDist.Zero, aircraft.Info.CruiseAltitude - targetAltitude));
```

Comment:
Can `aircraft.Info.CruiseAltitude` and `targetAltitude` be very different here? If so, this will make it jump instantly between positions. Why not call `VerticalTakeOffOrLandTick`?